shared/libraries/config_protocol/include/config_protocol/config_client_input_port_impl.h

```diff
@@ -69,27 +71,34 @@ inline ErrCode ConfigClientInputPortImpl::connect(ISignal* signal)
         {
             if (!this->deserializationComplete)
                 return Super::connect(signal);
-
             const auto signalPtr = SignalPtr::Borrow(signal);
+            if (!isSignalFromTheSameComponentTree(signalPtr))
+                return OPENDAQ_ERR_SIGNAL_NOT_ACCEPTED;
             {
                 std::scoped_lock lock(this->sync);
 
-                if (isConnected(signalPtr))
+                const auto connectedSignal = getConnectedSignal();
+                if (connectedSignal == signalPtr)
                     return OPENDAQ_IGNORED;
+                if (connectedSignal.assigned() && !clientComm->isComponentNested(connectedSignal.getGlobalId()))
+                    clientComm->disconnectExternalSignalFromServerInputPort(connectedSignal, remoteGlobalId);
             }
 
             const auto configObject = signalPtr.asPtrOrNull<IConfigClientObject>(true);
-            if (!configObject.assigned())
-                throw InvalidParameterException("Not a remote signal");
-
-             // TODO check that signal actually belongs to this device
+            if (configObject.assigned() && clientComm->isComponentNested(signalPtr.getGlobalId()))
+            {
+                StringPtr signalRemoteGlobalId;
+                checkErrorInfo(configObject->getRemoteGlobalId(&signalRemoteGlobalId));
 
-            StringPtr signalRemoteGlobalId;
-            checkErrorInfo(configObject->getRemoteGlobalId(&signalRemoteGlobalId));
+                auto params = ParamsDict({{"SignalId", signalRemoteGlobalId}});
 
-            auto params = ParamsDict({{"SignalId", signalRemoteGlobalId}});
+                clientComm->sendComponentCommand(remoteGlobalId, "ConnectSignal", params, nullptr);
+            }
+            else
+            {
+                clientComm->connectExternalSignalToServerInputPort(signalPtr, remoteGlobalId);
```

Comment:
We should probably check the protocol version before trying this, as this will not work on older servers.

Reply:
A check has been added to prevent execution on older server versions and reporting a more informative error to the user.